cirq/protocols/inverse.py

```diff
@@ -12,16 +12,54 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Any
+from typing import Any, overload, TYPE_CHECKING, TypeVar, Union
 
 import collections
 
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    import cirq
 
 # This is a special indicator value used by the inverse method to determine
 # whether or not the caller provided a 'default' argument.
 RaiseTypeErrorIfNotProvided = ([],)
 
 
+TDefault = TypeVar('TDefault')
+
+
+# pylint: disable=function-redefined
```

Comment:
I haven't done that much python with type checking.  Is this a common thing to do when a function can have multiple type inputs?  It looks weird.  Maybe it deserves a comment to explain what the purpose of all these stub functions are for?

Reply:
Ack. This is a standard feature of mypy. If you search `python @overload` or similar you pretty quickly come across what it means. This happens in many places in the code base, and it's not really practical to explain it in every place.